support full set of redis hash commands #20

72squared opened the pull request:
implemented all the hash commands.

I did not support streaming with hscan, we can leave that for a future implementation. Want to use reactive streams pattern.

Commit:
support full set of redis hash commands #20

scoquelin commented on modules/api/src/main/scala/com/github/scoquelin/arugula/commands/RedisHashAsyncCommands.scala:
```diff
@@ -1,19 +1,165 @@
+   * @param key The key
+   * @return A list of all the keys in the hash
+   */
+  def hKeys(key: K): Future[Seq[K]]
```
TLDR : I would recommend having a more fine grained type in the interface like `List`, because `Seq` is simply too generic and can lead to weird corner cases.

Explanation: Here we are controlling the implementation but let's say we choose to implement it as a `Stream` (which will conform to `Seq`), if the caller (who does not know the underlying implementation details) of the API calls `toList`, the call might never terminate when dealing with infinite collections...

scoquelin commented on modules/api/src/main/scala/com/github/scoquelin/arugula/commands/RedisHashAsyncCommands.scala:
```diff
@@ -1,19 +1,165 @@
+   * @param key The key
+   * @return A list of all the values in the hash
+   */
+  def hVals(key: K): Future[Seq[V]]
```
Same here

72squared commented:
I think I uncovered a Lettuce bug while testing. If you call ` hrandfieldWithvalues(K key, long count)`  on a key with fewer elements in it than the count, there is a nasty low-level redis parse error. I changed my test to avoid that case, but we might want to report it as an issue.

72squared commented:
converted all the responses to List instead of Seq. I also changed some of the inputs to List but perhaps that is a step too far? Accepting Seq or Iterable as an input is probably acceptable. I could change those input changes back to Seq or Iterable even? 

https://en.wikipedia.org/wiki/Robustness_principle

scoquelin commented on modules/api/src/main/scala/com/github/scoquelin/arugula/commands/RedisHashAsyncCommands.scala:
```diff
@@ -1,19 +1,165 @@
+   */
   def hSet(key: K, field: K, value: V): Future[Boolean]
+
+  /**
+   * Get the values of multiple hash fields
+   * @param key The key
+   * @param fields The fields
+   * @return A map of field -> value for each field that exists or None if the field does not exist
+   */
+  def hMGet(key: K, fields: K*): Future[ListMap[K, Option[V]]]
+
+  /**
+   * Get the values of multiple hash fields
+   * @param key The key
+   * @param fields The fields
+   * @return A map of field -> value for each field that exists or None if the field does not exist
+   */
+  def hMGet(key: K, fields: => List[K]): Future[Map[K, Option[V]]]
```
Should we keep both methods since they seem to achieve the same operation ?

Could we do maybe `def hMGet(key: K, fields: K*): Future[Map[K, Option[V]]]` only so that it's more in line with other operations ?

72squared replied:
I  personally hate the `fields: K*` approach since it prevents adding additional parameters later. Since that is unlikely to happen though, yeah I can get rid of the `def hMGet(key: K, fields: => List[K])` signature.

72squared replied:
okay, I removed the `def hMGet(key: K, fields: => List[K])` signature and left the other one. That should resolve the open issues.

scoquelin reviewed (approved):
Thanks for pushing hard on the hash commands! 🚀
